src/helper.cc

```diff
@@ -1201,7 +1203,7 @@ helperStatefulHandleRead(const Comm::ConnectionPointer &conn, char *, size_t len
         if (r && cbdataReferenceValid(r->request.data)) {
             r->reply.finalize();
             r->reply.reservationId = srv->reservationId;
-            r->request.callback(r->request.data, r->reply);
+            hlp->callBack(r);
         } else {
             debugs(84, DBG_IMPORTANT, "StatefulHandleRead: no callback data registered");
             called = 0;
```

Comment:
A few lines above this one:

```C++
        srv->requests.pop_front(); // we already have it in 'r'
```

When `r` is nil, we are hitting an std::list::pop_front() [UB](https://en.cppreference.com/w/cpp/container/list/pop_front) here: "If there are no elements in the container, the behavior is undefined."

I suggest making another prerequisite PR where we remove the special "no r" case by calling closePipesSafely() when the helper responds without being spoken to. There is similar code when we detect an overflowing buffer.

Besides complicating this refactoring and creating a UB in the official code, ignoring random content from helper has security implications because the affected helper may "smuggle" wrong response when requests do get queued after it responded but before we starting reading that response. Killing a helper process is the best reaction AFAICT (even if the situation is created by Squid bugs and even if some affected helpers are not buggy and "work" fine otherwise).

Reply:
> When r is nil,

I think the problem is that this function lacks std::list::empty() check before calling front() (all other similar cases have one). So r cannot be nil (if the list is not empty, front() cannot return nil). I posted this fix in a separate [PR231](https://github.com/measurement-factory/squid/pull/231).